api/layer/multipart_upload.go

```diff
@@ -595,9 +596,83 @@ func (n *layer) reUploadFollowingParts(ctx context.Context, uploadParams UploadP
 	for _, part := range parts {
 		uploadParams.PartNumber = part.Number
 
-		if err = n.reUploadPart(ctx, uploadParams, part.OID, bktInfo, multipartInfo); err != nil {
-			return fmt.Errorf("reupload number=%d: %w", part.Number, err)
+		if len(part.Elements) > 0 {
+			if err = n.reUploadSegmentedPart(ctx, uploadParams, part, bktInfo, multipartInfo); err != nil {
+				return fmt.Errorf("reupload number=%d: %w", part.Number, err)
+			}
+		} else {
+			if err = n.reUploadPart(ctx, uploadParams, part.OID, bktInfo, multipartInfo); err != nil {
+				return fmt.Errorf("reupload number=%d: %w", part.Number, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (n *layer) reUploadSegmentedPart(ctx context.Context, uploadParams UploadPartParams, part *data.PartInfo, bktInfo *data.BucketInfo, multipartInfo *data.MultipartInfo) error {
+	var (
+		eg                     errgroup.Group
+		pipeReader, pipeWriter = io.Pipe()
+	)
+
+	eg.Go(func() error {
+		for _, element := range part.Elements {
+			elementObj, err := n.objectGet(ctx, bktInfo, element.OID)
```

Comment:
I think it's overcomplicated. If you have a proper object, you may not care about its individual elements any more, you can just issue a GET for it, let node assemble individual parts, get an io.Reader and do io.Copy from it into larger object io.Writer.

Reply:
It is a key. If the part has elements, we slice it manually. This means we don't have a proper object and have to do all these actions